src/spa/src/query_evaluator/pql.cpp

```diff
@@ -0,0 +1,73 @@
+#include "pql.h"
+
+#include <map>
+#include <iostream>
+#include <algorithm>
+
+using namespace QE;
+
+Query::Query() 
+    : declarations(nullptr)
+    , selected_declaration(nullptr)
+    , such_that(nullptr)
+    , pattern(nullptr){};
+
+Query::Query(std::vector<QE::Declaration>* decls, QE::Declaration* selected_decl)
+    : declarations(decls)
+    , selected_declaration(selected_decl)
+    , such_that(nullptr)
+    , pattern(nullptr){};
+
+Query::~Query() {
+    delete declarations;
+    delete selected_declaration;
+    delete such_that;
+    delete pattern;
+}
+
+std::unique_ptr<Query> Query::makeStubQuery() {
+    auto decls = new std::vector<QE::Declaration>();
+    auto selected_decl = new QE::Declaration(DesignEntity::ASSIGN, "p");
+    return std::make_unique<Query>(decls, selected_decl);
+}
+
+bool Declaration::isValid() {
+    return std::regex_match(synonym, synonym_regex);
+}
+
+namespace QE {
+    // Hard to use unordered_map even though faster
+    // - need to define specialized hash for enum class
+    std::map<DesignEntity, std::string> designEntityToStringMap({
```

Comment:
Tempted to use http://aantron.github.io/better-enums/, we have this use case in many places.

Reply:
would also completely eliminate the need for the bottom map